README.md
# [torequests](https://github.com/ClericPy/torequests) [![PyPI](https://img.shields.io/pypi/v/torequests?style=plastic)](https://pypi.org/project/torequests/)[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/clericpy/torequests/Python%20package?style=plastic)](https://github.com/ClericPy/torequests/actions?query=workflow%3A%22Python+package%22)![PyPI - Wheel](https://img.shields.io/pypi/wheel/torequests?style=plastic)![PyPI - Python Version](https://img.shields.io/pypi/pyversions/torequests?style=plastic)![PyPI - Downloads](https://img.shields.io/pypi/dm/torequests?style=plastic)![PyPI - License](https://img.shields.io/pypi/l/torequests?style=plastic)

<!-- [![Downloads](https://pepy.tech/badge/torequests)](https://pepy.tech/project/torequests) -->

Briefly speaking, requests & aiohttp wrapper for asynchronous programming rookie, to shorten the code quantity. 

## Install:

> pip install torequests -U

#### requirements

##### python2.7 / python3.6+

    | requests			# python
    | futures 			# python2
    | aiohttp >= 3.0.5 	# python3
    | uvloop  			# python3

**optional:**

    | psutil
    | pyperclip

## Features

Inspired by [tomorrow](https://github.com/madisonmay/Tomorrow), to make async-coding brief & smooth, compatible for win32 / python 2&3.

* Convert any funtions into async-mode with concurrent.futures.
* Wrap requests lib with concurrent.futures to enjoy the concurrent performance.
* Simplify aiohttp, make it `requests-like`.
* Add FailureException to check the context of request failure.
* Add frequency control, prevent from anti-spider based on frequency check.
* Add retry for request.
* Alenty of common crawler utils.
* Compatible with [uniparser]( https://github.com/ClericPy/uniparser )

## Getting started

### 1. Async, threads - make functions asynchronous

```python
from torequests import threads, Async
import time


@threads(5)
def test1(n):
    time.sleep(n)
    return 'test1 ok'


def test2(n):
    time.sleep(n)
    return 'test1 ok'


start = int(time.time())
# here async_test2 is same as test1
async_test2 = Async(test2)
future = test1(1)
# future run in non blocking thread pool
print(future, ', %s s passed' % (int(time.time() - start)))
# call future.x will block main thread and get the future.result()
print(future.x, ', %s s passed' % (int(time.time() - start)))
# output:
# <NewFuture at 0x34b1d30 state=running> , 0 s passed
# test1 ok , 1 s passed
```
### 2. tPool - thread pool for async-requests

```python
from torequests.main import tPool
from torequests.logs import print_info
from torequests.utils import ttime

req = tPool()
test_url = 'http://p.3.cn'


def callback(task):
    return (len(task.content),
            print_info(
                ttime(task.task_start_time), '-> ', ttime(task.task_end_time),
                ',', round(task.task_cost_time * 1000), 'ms'))


ss = [req.get(test_url, retry=2, callback=callback) for i in range(3)]
# or [i.x for i in ss]
req.x
# task.cx returns the callback_result
ss = [i.cx for i in ss]
print_info(ss)
# [2020-01-21 16:56:23] temp_code.py(11): 2020-01-21 16:56:23 ->  2020-01-21 16:56:23 , 54 ms
# [2020-01-21 16:56:23] temp_code.py(11): 2020-01-21 16:56:23 ->  2020-01-21 16:56:23 , 55 ms
# [2020-01-21 16:56:23] temp_code.py(11): 2020-01-21 16:56:23 ->  2020-01-21 16:56:23 , 57 ms
# [2020-01-21 16:56:23] temp_code.py(18): [(612, None), (612, None), (612, None)]

```

#### 2.1 Performance.

```verilog
[3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 14:57:15) [MSC v.1915 64 bit (AMD64)]]: 2000 / 2000, 100.0%, cost 4.2121 seconds, 475.0 qps.
[2.7.15 (v2.7.15:ca079a3ea3, Apr 30 2018, 16:30:26) [MSC v.1500 64 bit (AMD64)]]: 2000 / 2000, 100%, cost 9.4462 seconds, 212.0 qps.
```

```python
import timeit
from torequests import tPool
import sys

req = tPool()
start_time = timeit.default_timer()
oks = 0
total = 2000
# concurrent all the tasks
tasks = [req.get('http://127.0.0.1:8080') for num in range(total)]
for task in tasks:
    r = task.x
    if r.text == 'ok':
        oks += 1
end_time = timeit.default_timer()
succ_rate = oks * 100 / total
cost_time = round(end_time - start_time, 4)
version = sys.version
qps = round(total / cost_time, 0)
print(
    '[{version}]: {oks} / {total}, {succ_rate}%, cost {cost_time} seconds, {qps} qps.'
    .format(**vars()))
```

### 3. Requests - aiohttp-wrapper

```python
from torequests.dummy import Requests
from torequests.logs import print_info
from torequests.utils import ttime
req = Requests(frequencies={'p.3.cn': (2, 1)})


def callback(task):
    return (len(task.content),
            print_info(
                ttime(task.task_start_time), '->', ttime(task.task_end_time),
                ',', round(task.task_cost_time * 1000), 'ms'))


ss = [
    req.get('http://p.3.cn', retry=1, timeout=5, callback=callback)
    for i in range(4)
]
req.x  # this line can be removed
ss = [i.cx for i in ss]
print_info(ss)
# [2020-01-21 18:15:33] temp_code.py(11): 2020-01-21 18:15:32 -> 2020-01-21 18:15:33 , 1060 ms
# [2020-01-21 18:15:33] temp_code.py(11): 2020-01-21 18:15:32 -> 2020-01-21 18:15:33 , 1061 ms
# [2020-01-21 18:15:34] temp_code.py(11): 2020-01-21 18:15:32 -> 2020-01-21 18:15:34 , 2081 ms
# [2020-01-21 18:15:34] temp_code.py(11): 2020-01-21 18:15:32 -> 2020-01-21 18:15:34 , 2081 ms
# [2020-01-21 18:15:34] temp_code.py(20): [(612, None), (612, None), (612, None), (612, None)]
```

#### 3.1 Performance.

>  aiohttp is almostly 3 times faster than requests + ThreadPoolExecutor, even without uvloop on windows10.

```verilog
3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 14:57:15) [MSC v.1915 64 bit (AMD64)]
sync_test: 2000 / 2000, 100.0%, cost 1.2965 seconds, 1543.0 qps.
async_test: 2000 / 2000, 100.0%, cost 1.2834 seconds, 1558.0 qps.

Sync usage has little performance lost.
```

```python
import asyncio
import sys
import timeit

from torequests.dummy import Requests


def sync_test():
    with Requests() as req:
        start_time = timeit.default_timer()
        oks = 0
        total = 2000
        # concurrent all the tasks
        tasks = [req.get('http://127.0.0.1:8080') for num in range(total)]
        req.x
        for task in tasks:
            if task.text == 'ok':
                oks += 1
        end_time = timeit.default_timer()
        succ_rate = oks * 100 / total
        cost_time = round(end_time - start_time, 4)
        qps = round(total / cost_time, 0)
        print(
            f'sync_test: {oks} / {total}, {succ_rate}%, cost {cost_time} seconds, {qps} qps.'
        )


async def async_test():
    req = Requests()
    async with Requests() as req:
        start_time = timeit.default_timer()
        oks = 0
        total = 2000
        # concurrent all the tasks
        tasks = [req.get('http://127.0.0.1:8080') for num in range(total)]
        for task in tasks:
            r = await task
            if r.text == 'ok':
                oks += 1
        end_time = timeit.default_timer()
        succ_rate = oks * 100 / total
        cost_time = round(end_time - start_time, 4)
        qps = round(total / cost_time, 0)
        print(
            f'async_test: {oks} / {total}, {succ_rate}%, cost {cost_time} seconds, {qps} qps.'
        )


if __name__ == "__main__":
    print(sys.version)
    sync_test()
    loop = asyncio.get_event_loop()
    loop.run_until_complete(async_test())

```

#### 3.2 using torequests.dummy.Requests in async environment.

```python
import asyncio

import uvicorn
from starlette.applications import Starlette
from starlette.responses import PlainTextResponse
from torequests.dummy import Requests

api = Starlette()
api.req = Requests()


@api.route('/')
async def index(req):
    # await for Response or FailureException
    r = await api.req.get('http://p.3.cn', timeout=(1, 1))
    return PlainTextResponse(r.content)


if __name__ == "__main__":
    uvicorn.run(api)

```

#### 3.3 using torequests.aiohttp_dummy.Requests for better performance.

> Removes the frequency_controller & sync usage (task.x) & compatible args of requests for good performance, but remains retry / callback / referer_info.

```python
# -*- coding: utf-8 -*-

from torequests.aiohttp_dummy import Requests
from asyncio import get_event_loop


async def main():
    url = 'http://httpbin.org/json'
    # simple usage, catch_exception defaults to True
    async with Requests() as req:
        r = await req.get(url, retry=1, encoding='u8', timeout=3)
        print(r.ok)
        print(r.url)
        print(r.content[:5])
        print(r.status_code)
        print(r.text[:5])
        print(r.json()['slideshow']['title'])
        print(r.is_redirect)
        print(r.encoding)
    # or use `req = Requests()` without context
    req = Requests()
    r = await req.get(
        url,
        referer_info=123,
        callback=lambda r: r.referer_info + 1,
    )
    print(r)  # 124


get_event_loop().run_until_complete(main())

```


### 4. utils: some useful crawler toolkits

    ClipboardWatcher: watch your clipboard changing.
    Counts: counter while every time being called.
    Null: will return self when be called, and alway be False.
    Regex: Regex Mapper for string -> regex -> object.
    Saver: simple object persistent toolkit with pickle/json.
    Timer: timing tool.
    UA: some common User-Agents for crawler.
    curlparse: translate curl-string into dict of request.
    md5: str(obj) -> md5_string.
    print_mem: show the proc-mem-cost with psutil, use this only for lazinesssss.
    ptime: %Y-%m-%d %H:%M:%S -> timestamp.
    ttime: timestamp -> %Y-%m-%d %H:%M:%S
    slice_by_size: slice a sequence into chunks, return as a generation of chunks with size.
    slice_into_pieces: slice a sequence into n pieces, return a generation of n pieces.
    timeago: show the seconds as human-readable.
    unique: unique one sequence.
    find_one: use regex like Javascript to find one string with index(like [0], [1]).
    ...


## Benchmarks
> Benchmark of concurrent is not very necessary and accurate, just take a look.

### Test Server: golang net/http

Source code: [go_test_server.go](https://github.com/ClericPy/torequests/blob/master/benchmarks/go_test_server.go)

### Client Testing Code

Source code: [py_test_client.py](https://github.com/ClericPy/torequests/blob/master/benchmarks/py_test_client.py)

### Test Result on Windows (without uvloop)

> Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz (12 CPUs), ~2.2GHz; 12 logical CPUs.

```verilog
Windows-10-10.0.18362-SP0
3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 14:57:15) [MSC v.1915 64 bit (AMD64)]
['aiohttp(3.6.2)', 'torequests(4.9.14)', 'requests(2.23.0)', 'httpx(0.12.1)']
================================================================================
test_aiohttp             : 2000 / 2000 = 100.0%, cost 1.289s, 1551 qps, 100% standard.
test_dummy               : 2000 / 2000 = 100.0%, cost 1.397s, 1432 qps,  92% standard.
test_aiohttp_dummy       : 2000 / 2000 = 100.0%, cost 1.341s, 1491 qps,  96% standard.
test_httpx               : 2000 / 2000 = 100.0%, cost 4.065s,  492 qps,  32% standard.
test_tPool               : 2000 / 2000 = 100.0%, cost 5.161s,  388 qps,  25% standard.
```

### Test Result on Linux (with uvloop)

> Intel(R) Xeon(R) Platinum 8163 CPU @ 2.50GHz; 1 logical CPU.

```verilog
Linux-4.15.0-13-generic-x86_64-with-Ubuntu-18.04-bionic
3.7.3 (default, Apr  3 2019, 19:16:38)
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
['aiohttp(3.6.2)', 'torequests(4.9.14)', 'requests(2.23.0)', 'httpx(0.12.1)']
================================================================================
test_aiohttp             : 2000 / 2000 = 100.0%, cost 0.652s, 3068 qps, 100% standard.

changelog.md

---

later (4.9.10+) changes view: https://github.com/ClericPy/torequests/releases

- 4.9.9
    - fix catch_exception error

- 4.9.6 ~ 4.9.8
    - fix python3.8 contend generator
    - use asend instead of async for
    - fix default_host_frequency strategy
    - new faster and tiny frequency controller

- 4.9.5
    - fix aiohttp verify param compatible

- 4.9.4
    - fix aiohttp 4 Compatible issue
    - update readme and benchmark

- 4.9.3
    - hotfix setup.py

- 4.9.2
    - fix bugs

- 4.9.1
    - add new Frequency implementation
    - fix python2 queue import error
    - remove useless parsers, uniparser is the new choice
    - update benchmarks
    - cancel frequency put_tasks if `del`

- 4.9.0
    - fix dummy Requests task.cx not return callback result #36
    - add `__aenter__` for dummy
    - fix frequency
    - fix task_cost_time

- 4.8.21
    - remove nonsense semphore for performance lost
    - add some extras_requires

- 4.8.20
    - fix runtime error for running loop

- 4.8.19
    - Fix curlparse encoding error
    - add test 3.8
    - fix aiohttp unclosed connection

- 4.8.18
    - hotfix curlparse issue by backslash string.

- 4.8.17
    - hotfix version crash of 4.8.16

- 4.8.16 - [deleted]
    - upgrade aiohttp 3.6.2 for deprecation warning

- 4.8.15
    - update setup.py requirements

- 4.8.14
    - use ssl args for newest aiohttp

- 4.8.13
    - fix dummy unclosed connection

- 4.8.12
    - fix utils.curlparse post form encoding issue

- 4.8.10
    - fix dummy.Requests "Unclosed connector"
    - add sort_url_query

- 4.8.9
    - fix issue: close aiohttp.ClientSession as coroutine

- 4.8.8
    - remove useless arg of torequests.dummy.Requests: reuse_running_loop

- 4.8.7
    - support tuple type timeout for dummy, like
        - `req.get(url, timeout=(1, 3))`

- 4.8.6 and before
    - ignore


torequests/main.py
#! coding:utf-8
# python2 requires: pip install futures

import atexit
from concurrent.futures import (ProcessPoolExecutor, ThreadPoolExecutor,
                                as_completed)
from concurrent.futures._base import (
    CANCELLED, CANCELLED_AND_NOTIFIED, FINISHED, PENDING, RUNNING,
    CancelledError, Error, Executor, Future, TimeoutError)
from concurrent.futures.thread import _threads_queues, _WorkItem
from functools import wraps
from logging import getLogger
from threading import Timer
from time import time as time_time
from weakref import WeakSet

from requests import PreparedRequest, RequestException, Session
from requests.adapters import HTTPAdapter
from urllib3 import disable_warnings

from .configs import Config
from .exceptions import FailureException
from .frequency_controller.sync_tools import Frequency
from .versions import PY2, PY3

try:
    from queue import Queue
except ImportError:
    from Queue import Queue
if PY3:
    from concurrent.futures.process import BrokenProcessPool

__all__ = [
    "Pool", "ProcessPool", "NewFuture", "Async", "threads",
    "get_results_generator", "run_after_async", "tPool", "get", "post",
    "options", "delete", "put", "head", "patch", "request", "disable_warnings"
]
logger = getLogger("torequests")


def _abandon_all_tasks():
    """Only used for abandon_all_tasks and exit the main thread,
    to prevent the main thread waiting for unclosed thread while exiting."""
    _threads_queues.clear()


def ensure_waiting_for_threads():
    if Config.wait_futures_before_exiting:
        _abandon_all_tasks()


atexit.register(ensure_waiting_for_threads)


class NewExecutorPoolMixin(Executor):
    """Add async_func decorator for wrapping a function to return the NewFuture."""

    def async_func(self, function):
        """Decorator for let a normal function return the NewFuture"""

        @wraps(function)
        def wrapped(*args, **kwargs):
            return self.submit(function, *args, **kwargs)

        return wrapped

    def close(self, wait=True):
        """Same as self.shutdown"""
        return self.shutdown(wait=wait)

    def _get_cpu_count(self):
        """Get the cpu count."""
        try:
            from multiprocessing import cpu_count

            return cpu_count()
        except Exception as e:
            logger.error("_get_cpu_count failed for %s" % e)

    @property
    def x(self):
        """Return self.wait_futures_done"""
        return self.wait_futures_done(list(self._all_futures))

    def wait_futures_done(self, tasks=None):
        # ignore the order of tasks
        tasks = tasks or self._all_futures
        fs = []
        try:
            for f in as_completed(tasks, timeout=self._timeout):
                fs.append(f.x)
        except TimeoutError:
            pass
        return fs


class Pool(ThreadPoolExecutor, NewExecutorPoolMixin):
    """Let ThreadPoolExecutor use NewFuture instead of origin concurrent.futures.Future.

    WARNING: NewFutures in Pool will not block main thread without NewFuture.x.

    Basic Usage::

            from torequests.main import Pool
            import time

            pool = Pool()


            def use_submit(i):
                time.sleep(i)
                result = 'use_submit: %s' % i
                print(result)
                return result


            @pool.async_func
            def use_decorator(i):
                time.sleep(i)
                result = 'use_decorator: %s' % i
                print(result)
                return result


            tasks = [pool.submit(use_submit, i) for i in (2, 1, 0)
                    ] + [use_decorator(i) for i in (2, 1, 0)]
            # pool.x can be ignore
            pool.x
            results = [i.x for i in tasks]
            print(results)

            # use_submit: 0
            # use_decorator: 0
            # use_submit: 1
            # use_decorator: 1
            # use_submit: 2
            # use_decorator: 2
            # ['use_submit: 2', 'use_submit: 1', 'use_submit: 0', 'use_decorator: 2', 'use_decorator: 1', 'use_decorator: 0']
    """

    def __init__(self,
                 n=None,
                 timeout=None,
                 default_callback=None,
                 catch_exception=True,
                 *args,
                 **kwargs):
        n = n or kwargs.pop("max_workers", None)
        if PY2 and n is None:
            # python2 n!=None
            n = (self._get_cpu_count() or 1) * 5
        super(Pool, self).__init__(n, *args, **kwargs)
        #: set the default timeout
        self._timeout = timeout
        #: set the default_callback if not set single task's callback
        self.default_callback = default_callback
        #: WeakSet of _all_futures for self.x
        self._all_futures = WeakSet()
        #: catch_exception=True will not raise exceptions, return object FailureException(exception)
        self.catch_exception = catch_exception

    @property
    def all_tasks(self):
        """Keep the same api for dummy, return self._all_futures actually"""
        return self._all_futures

    def submit(self, func, *args, **kwargs):
        """Submit a function to the pool, `self.submit(function,arg1,arg2,arg3=3)`"""

        with self._shutdown_lock:
            if self._shutdown:
                raise RuntimeError("cannot schedule new futures after shutdown")
            callback = kwargs.pop("callback", self.default_callback)
            future = NewFuture(
                self._timeout,
                args,
                kwargs,
                callback=callback,
                catch_exception=self.catch_exception,
            )
            w = _WorkItem(future, func, args, kwargs)
            self._work_queue.put(w)
            self._adjust_thread_count()
            self._all_futures.add(future)
            return future


class ProcessPool(ProcessPoolExecutor, NewExecutorPoolMixin):
    """Simple ProcessPool covered ProcessPoolExecutor.
    ::

        from torequests.main import ProcessPool
        import time

        pool = ProcessPool()


        def use_submit(i):
            time.sleep(i)
            result = 'use_submit: %s' % i
            print(result)
            return result


        def main():
            tasks = [pool.submit(use_submit, i) for i in (2, 1, 0)]
            # pool.x can be ignore
            pool.x
            results = [i.x for i in tasks]
            print(results)


        if __name__ == '__main__':
            main()

        # ['use_submit: 2', 'use_submit: 1', 'use_submit: 0']
        # use_submit: 0
        # use_submit: 1
        # use_submit: 2
    """

    def __init__(self,
                 n=None,
                 timeout=None,
                 default_callback=None,
                 catch_exception=True,
                 *args,
                 **kwargs):
        n = n or kwargs.pop("max_workers", None)
        if PY2 and n is None:
            # python2 n!=None
            n = self._get_cpu_count() or 1
        super(ProcessPool, self).__init__(n, *args, **kwargs)
        self._timeout = timeout
        self.default_callback = default_callback
        self._all_futures = WeakSet()
        self.catch_exception = catch_exception

    def submit(self, func, *args, **kwargs):
        """Submit a function to the pool, `self.submit(function,arg1,arg2,arg3=3)`"""

        with self._shutdown_lock:
            if PY3 and self._broken:
                raise BrokenProcessPool(
                    "A child process terminated "
                    "abruptly, the process pool is not usable anymore")
            if self._shutdown_thread:
                raise RuntimeError("cannot schedule new futures after shutdown")
            callback = kwargs.pop("callback", self.default_callback)
            future = NewFuture(
                self._timeout,
                args,
                kwargs,
                callback=callback,
                catch_exception=self.catch_exception,
            )
            w = _WorkItem(future, func, args, kwargs)
            self._pending_work_items[self._queue_count] = w
            self._work_ids.put(self._queue_count)
            self._queue_count += 1
            self._result_queue.put(None)
            self._start_queue_management_thread()
            if PY2:
                self._adjust_process_count()
            self._all_futures.add(future)
            return future

    def async_func(self, *args):
        """Decorator mode not support for ProcessPool for _pickle.PicklingError."""
        raise NotImplementedError


class NewFuture(Future):
    """Add `.x` attribute and timeout args for original Future class

    WARNING: Future thread will not stop running until function finished or pid killed.

    :attr cx: blocking until the task finish and return the callback_result.
    :attr x: blocking until the task finish and return the value as `coro` returned.
    :attr task_start_time: timestamp when the task start up.
    :attr task_end_time: timestamp when the task end up.
    :attr task_cost_time: seconds of task costs.
    :param catch_exception: `True` will catch all exceptions and return as :class:`FailureException <FailureException>`
    """

    if PY3:
        from ._py3_patch import _new_future_await

        __await__ = _new_future_await

    def __init__(self,
                 timeout=None,
                 args=None,
                 kwargs=None,
                 callback=None,
                 catch_exception=True):
        super(NewFuture, self).__init__()
        self._timeout = timeout
        self._args = args or ()
        self._kwargs = kwargs or {}
        self._callback_result = None
        self.catch_exception = catch_exception
        self.task_start_time = time_time()
        self.task_end_time = 0
        self.task_cost_time = 0
        self._user_callbacks = set()
        if callback:
            if not isinstance(callback, (list, tuple)):
                callback = [callback]
            for fn in callback:
                self.add_done_callback(fn)
                self._user_callbacks.add(fn)

    def __getattr__(self, name):
        return getattr(self.x, name)

    def _invoke_callbacks(self):
        """Record the task_end_time & task_cost_time, set result for self._callback_result."""
        self.task_end_time = time_time()
        self.task_cost_time = self.task_end_time - self.task_start_time
        with self._condition:
            for callback in self._done_callbacks:
                try:
                    result = callback(self)
                    if callback in self._user_callbacks:
                        self._callback_result = result
                except Exception as e:
                    logger.error("exception calling callback for %s" % e)
            self._condition.notify_all()

    @property
    def _callbacks(self):
        """Keep same api for NewTask."""
        return self._done_callbacks

    @property
    def cx(self):
        """Block the main thead until future finish, return the future.callback_result."""
        return self.callback_result

    @property
    def callback_result(self):
        """Block the main thead until future finish, return the future.callback_result."""
        if self._state in [PENDING, RUNNING]:
            self.x
        if self._user_callbacks:
            return self._callback_result
        else:
            return self.x

    @property
    def x(self):
        """Block the main thead until future finish, return the future.result()."""
        with self._condition:
            result = None
            if not self.done():
                self._condition.wait(self._timeout)
            if not self.done():
                # timeout
                self.set_exception(TimeoutError())
            if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:
                # cancelled
                result = CancelledError()
            elif self._state == FINISHED:
                # finished
                if self._exception:
                    result = self._exception
                else:
                    result = self._result
            if isinstance(result, Exception):
                if self.catch_exception:
                    result = FailureException(result)
                    return result
                else:
                    raise result
            return result


def Async(f, n=None, timeout=None):
    """Concise usage for pool.submit.

    Basic Usage Asnyc & threads ::

        from torequests.main import Async, threads
        import time


        def use_submit(i):
            time.sleep(i)
            result = 'use_submit: %s' % i
            print(result)
            return result


        @threads()
        def use_decorator(i):
            time.sleep(i)
            result = 'use_decorator: %s' % i
            print(result)
            return result


        new_use_submit = Async(use_submit)
        tasks = [new_use_submit(i) for i in (2, 1, 0)
                ] + [use_decorator(i) for i in (2, 1, 0)]
        print([type(i) for i in tasks])
        results = [i.x for i in tasks]
        print(results)

        # use_submit: 0
        # use_decorator: 0
        # [<class 'torequests.main.NewFuture'>, <class 'torequests.main.NewFuture'>, <class 'torequests.main.NewFuture'>, <class 'torequests.main.NewFuture'>, <class 'torequests.main.NewFuture'>, <class 'torequests.main.NewFuture'>]
        # use_submit: 1
        # use_decorator: 1
        # use_submit: 2
        # use_decorator: 2
        # ['use_submit: 2', 'use_submit: 1', 'use_submit: 0', 'use_decorator: 2', 'use_decorator: 1', 'use_decorator: 0']
    """
    return threads(n=n, timeout=timeout)(f)


def threads(n=None, timeout=None):
    """Decorator usage like Async."""
    return Pool(n, timeout).async_func


def get_results_generator(future_list, timeout=None, sort_by_completed=False):
    """Return as a generator of tasks order by completed sequence."""
    try:
        # python2 not support yield from
        if sort_by_completed:
            for future in as_completed(future_list, timeout=timeout):
                yield future.x
        else:
            for future in future_list:
                yield future.x
    except TimeoutError:
        return


def run_after_async(seconds, func, *args, **kwargs):
    """Run the function after seconds asynchronously."""
    t = Timer(seconds, func, args, kwargs)
    t.daemon = True
    t.start()
    return t


class FailedRequest(PreparedRequest):
    allow_keys = {
        "method",
        "url",
        "headers",
        "files",
        "data",
        "params",
        "auth",
        "cookies",
        "hooks",
        "json",
    }

    def __init__(self, **kwargs):
        # self.kwargs for retry tPool.request
        self.kwargs = kwargs
        filted_kwargs = {
            key: value
            for key, value in kwargs.items()
            if key in self.allow_keys
        }
        super(FailedRequest, self).__init__()
        self.prepare(**filted_kwargs)


class tPool(object):
    """Async wrapper for requests.

    :param n: thread pool size for concurrent limit.
    :param interval: time.sleep(interval) after each task finished.
    :param timeout: timeout for each task.result(timeout). But it will not shutdown the raw funtion.
    :param session: individually given a available requests.Session instance if necessary.
    :param catch_exception: `True` will catch all exceptions and return as :class:`FailureException <FailureException>`
    :param default_callback: default_callback for tasks which not set callback param.

    Usage::

        from torequests.main import tPool
        from torequests.logs import print_info

        trequests = tPool(2, 1)
        test_url = 'http://p.3.cn'
        ss = [
            trequests.get(
                test_url,
                retry=2,
                callback=lambda x: (len(x.content), print_info(len(x.content))))
            for i in range(3)
        ]
        # or [i.x for i in ss]
        trequests.x
        ss = [i.cx for i in ss]
        print_info(ss)

        # [2020-02-11 11:36:33] temp_code.py(10): 612
        # [2020-02-11 11:36:33] temp_code.py(10): 612
        # [2020-02-11 11:36:34] temp_code.py(10): 612
        # [2020-02-11 11:36:34] temp_code.py(16): [(612, None), (612, None), (612, None)]
    """

    def __init__(
            self,
            n=None,
            interval=0,
            timeout=None,
            session=None,
            catch_exception=True,
            default_callback=None,
    ):
        self.pool = Pool(n, timeout)
        self.session = session if session else Session()
        self.n = n or 10
        # adapt the concurrent limit.
        custom_adapter = HTTPAdapter(
            pool_connections=self.n, pool_maxsize=self.n)
        self.session.mount("http://", custom_adapter)
        self.session.mount("https://", custom_adapter)
        self.interval = interval
        self.catch_exception = catch_exception
        self.default_callback = default_callback
        self.frequency = Frequency(self.n, self.interval)

    @property
    def all_tasks(self):
        """Return self.pool._all_futures"""
        return self.pool._all_futures

    @property
    def x(self):
        """Return self.pool.x"""
        return self.pool.x

    def close(self, wait=False):
        """Close session, shutdown pool."""
        self.session.close()
        self.pool.shutdown(wait=wait)

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def __del__(self):
        self.close()

    def _request(self, method, url, retry=0, **kwargs):
        kwargs["url"] = url
        kwargs["method"] = method
        # non-official request args
        referer_info = kwargs.pop("referer_info", None)
        encoding = kwargs.pop("encoding", None)
        for _ in range(retry + 1):
            with self.frequency:
                try:
                    resp = self.session.request(**kwargs)
                    if encoding:
                        resp.encoding = encoding
                    logger.debug("%s done, %s" % (url, kwargs))
                    resp.referer_info = referer_info
                    return resp
                except (RequestException, Error) as e:
                    error = e
                    logger.debug(
                        "Retry %s for the %s time, Exception: %s . kwargs= %s" %
                        (url, _ + 1, e, kwargs))
                    continue
        # for unofficial request args
        kwargs["retry"] = retry
        if referer_info:
            kwargs["referer_info"] = referer_info
        if encoding:
            kwargs["encoding"] = encoding
        logger.debug("Retry %s & failed: %s." % (retry, error))
        failure = FailureException(error)
        failure.request = FailedRequest(**kwargs)
        if self.catch_exception:
            return failure
        else:
            raise failure

    def request(self, method, url, callback=None, retry=0, **kwargs):
        """Similar to `requests.request`, but return as NewFuture."""
        return self.pool.submit(
            self._request,
            method=method,
            url=url,
            retry=retry,
            callback=callback or self.default_callback,
            **kwargs)

    def get(self, url, params=None, callback=None, retry=0, **kwargs):
        """Similar to `requests.get`, but return as NewFuture."""
        kwargs.setdefault("allow_redirects", True)
        return self.request(
            "get",
            url=url,
            params=params,
            callback=callback,
            retry=retry,
            **kwargs)

    def post(self, url, data=None, json=None, callback=None, retry=0, **kwargs):
        """Similar to `requests.post`, but return as NewFuture."""
        return self.request(
            "post",
            url=url,
            data=data,
            json=json,
            callback=callback,
            retry=retry,
            **kwargs)

    def delete(self, url, callback=None, retry=0, **kwargs):
        """Similar to `requests.delete`, but return as NewFuture."""
        return self.request(
            "delete", url=url, callback=callback, retry=retry, **kwargs)

    def put(self, url, data=None, callback=None, retry=0, **kwargs):
        """Similar to `requests.put`, but return as NewFuture."""
        return self.request(
            "put", url=url, data=data, callback=callback, retry=retry, **kwargs)

    def head(self, url, callback=None, retry=0, **kwargs):
        """Similar to `requests.head`, but return as NewFuture."""
        kwargs.setdefault("allow_redirects", False)
        return self.request(
            "head", url=url, callback=callback, retry=retry, **kwargs)

    def options(self, url, callback=None, retry=0, **kwargs):
        """Similar to `requests.options`, but return as NewFuture."""
        kwargs.setdefault("allow_redirects", True)
        return self.request(
            "options", url=url, callback=callback, retry=retry, **kwargs)

    def patch(self, url, callback=None, retry=0, **kwargs):
        """Similar to `requests.patch`, but return as NewFuture."""
        return self.request(
            "patch", url=url, callback=callback, retry=retry, **kwargs)


def get(url, params=None, callback=None, retry=0, **kwargs):
    return tPool().get(
        url, params=params, callback=callback, retry=retry, **kwargs)


def post(url, data=None, json=None, callback=None, retry=0, **kwargs):
    return tPool().post(
        url, data=data, json=json, callback=callback, retry=retry, **kwargs)


def delete(url, callback=None, retry=0, **kwargs):
    return tPool().delete(url, callback=callback, retry=retry, **kwargs)


def put(url, data=None, callback=None, retry=0, **kwargs):
    return tPool().put(url, data=data, callback=callback, retry=retry, **kwargs)


def head(url, callback=None, retry=0, **kwargs):
    return tPool().head(url, callback=callback, retry=retry, **kwargs)


def options(url, callback=None, retry=0, **kwargs):
    return tPool().options(url, callback=callback, retry=retry, **kwargs)


def patch(url, callback=None, retry=0, **kwargs):
    return tPool().patch(url, callback=callback, retry=retry, **kwargs)


def request(method, url, callback=None, retry=0, **kwargs):
    return tPool().request(
        method, url, callback=callback, retry=retry, **kwargs)
